cpp/src/join/unique_hash_join.cu

```diff
@@ -0,0 +1,334 @@
+  auto const d_hasher   = row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
+
+  auto const iter = cudf::detail::make_counting_transform_iterator(
+    0, build_keys_fn<decltype(d_hasher), lhs_index_type>{d_hasher});
+
+  size_type const build_table_num_rows{build.num_rows()};
+  if (this->_nulls_equal == cudf::null_equality::EQUAL or (not cudf::nullable(this->_build))) {
+    this->_hash_table.insert_async(iter, iter + build_table_num_rows, stream.value());
+  } else {
+    auto stencil = thrust::counting_iterator<size_type>{0};
+    auto const row_bitmask =
+      cudf::detail::bitmask_and(this->_build, stream, rmm::mr::get_current_device_resource()).first;
+    auto const pred =
+      cudf::detail::row_is_valid{reinterpret_cast<bitmask_type const*>(row_bitmask.data())};
+
+    // insert valid rows
+    this->_hash_table.insert_if_async(
+      iter, iter + build_table_num_rows, stencil, pred, stream.value());
+  }
+}
+
+template <typename Hasher, cudf::has_nested HasNested>
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+unique_hash_join<Hasher, HasNested>::inner_join(std::optional<std::size_t> output_size,
+                                                rmm::cuda_stream_view stream,
+                                                rmm::mr::device_memory_resource* mr) const
+{
+  CUDF_FUNC_RANGE();
+
+  size_type const probe_table_num_rows{this->_probe.num_rows()};
+
+  std::size_t const join_size = output_size ? *output_size : probe_table_num_rows;
+
+  auto left_indices  = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+  auto right_indices = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+
+  auto const probe_row_hasher =
+    cudf::experimental::row::hash::row_hasher{this->_preprocessed_probe};
+  auto const d_probe_hasher = probe_row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
+  auto const iter           = cudf::detail::make_counting_transform_iterator(
+    0, build_keys_fn<decltype(d_probe_hasher), rhs_index_type>{d_probe_hasher});
+  auto counter = rmm::device_scalar<cudf::size_type>{stream};
+  counter.set_value_to_zero_async(stream);
+
+  cudf::detail::grid_1d grid{probe_table_num_rows, DEFAULT_JOIN_BLOCK_SIZE};
+  unique_join_probe_kernel<<<grid.num_blocks, grid.num_threads_per_block, 0, stream.value()>>>(
+    iter,
+    probe_table_num_rows,
+    this->_hash_table.ref(cuco::experimental::op::find),
+    left_indices->data(),
+    right_indices->data(),
+    counter.data());
+
+  auto const actual_size = counter.value(stream);
+  left_indices->resize(actual_size, stream);
+  right_indices->resize(actual_size, stream);
+
+  return {std::move(left_indices), std::move(right_indices)};
+}
+
+template <typename Hasher, cudf::has_nested HasNested>
+std::size_t unique_hash_join<Hasher, HasNested>::inner_join_size(rmm::cuda_stream_view stream) const
```

Comment:
Must a size estimate be implemented at all? Delete if possible.

Reply:
Deleted